base/applications/mspaint/mouse.cpp

```diff
@@ -277,293 +511,346 @@ endPaintingL(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
+        if (pointSP == 255)
+            pointSP--;
+    }
+    virtual void OnCancelDraw()
+    {
+        imageModel.ResetToPrevious();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_ELLIPSE
+struct EllipseTool : ToolBase
+{
+    EllipseTool() : ToolBase(TOOL_ELLIPSE)
+    {
+    }
+
+    virtual void OnDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Ellp(m_hdc, start.x, start.y, x, y, m_fg, m_bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+        else
+            Ellp(m_hdc, start.x, start.y, x, y, m_bg, m_fg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+    }
+    virtual void OnMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnMove(bLeftButton, x, y);
+    }
+    virtual void OnUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        ToolBase::OnUp(bLeftButton, x, y);
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+    }
+    virtual void OnCancelDraw()
+    {
+        OnUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_RRECT
+struct RRectTool : ToolBase
+{
+    RRectTool() : ToolBase(TOOL_RRECT)
+    {
+    }
+
+    virtual void OnDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            RRect(m_hdc, start.x, start.y, x, y, m_fg, m_bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+        else
+            RRect(m_hdc, start.x, start.y, x, y, m_bg, m_fg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+    }
+    virtual void OnMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnMove(bLeftButton, x, y);
+    }
+    virtual void OnUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        ToolBase::OnUp(bLeftButton, x, y);
+        imageModel.ResetToPrevious();
+        draw(bLeftButton, x, y);
+    }
+    virtual void OnCancelDraw()
+    {
+        OnUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+/*static*/ ToolBase*
+ToolBase::createToolObject(TOOLTYPE type)
+{
+    switch (type)
+    {
```

Comment:
Since we've always just one single tool active at any time... I'm wondering whether we could allow this game of allocations/deallocations, but instead have each individual tool being sort of a singleton that inherits from the base class, and then switch the tool around like that 🤔 (just assigning the current tool pointer to an existing object, and back... )

Reply:
Used cache for the tool objects.